benchmarks/Bio/BCR.jmd

```diff
@@ -90,24 +97,62 @@ test_sol  = TestSolution(sol)
 
 ## Setups
 
+#### Declare pre-conditioners
+```julia
+using IncompleteLU, LinearAlgebra
+
+jaccache = sparsejacprob.f.jac(oprob.u0,oprob.p,0.0)
+W = I - 1.0*jaccache
+prectmp = ilu(W, τ = 50.0)
+preccache = Ref(prectmp)
+
+τ1 = 1e2
+function psetupilu(p, t, u, du, jok, jcurPtr, gamma)
+    if jok
+        sparsejacprob.f.jac(jaccache,u,p,t)
+        jcurPtr[] = true
+
+        # W = I - gamma*J
+        @. W = -gamma*jaccache
+        idxs = diagind(W)
+        @. @view(W[idxs]) = @view(W[idxs]) + 1
+
+        # Build preconditioner on W
+        preccache[] = ilu(W, τ = τ1)
+    end
+end
+function precilu(z,r,p,t,y,fy,gamma,delta,lr)
+    ldiv!(z,preccache[],r)
+end
+
+τ2 = 1e10
+function get_julia_pcs(τ)
+    function incompletelu(W,du,u,p,t,newW,Plprev,Prprev,solverdata)
+        if newW === nothing || newW
+            Pl = ilu(convert(AbstractMatrix,W), τ = τ2)
+        else
+            Pl = Plprev
+        end
+        Pl,nothing
+    end
+    return incompletelu
+end;
+```
+
+#### Declare solvers
+
 ```julia
 abstols = 1.0 ./ 10.0 .^ (5:8)
 reltols = 1.0 ./ 10.0 .^ (5:8);
+
 setups = [
-          #Dict(:alg=>Rosenbrock23(autodiff=false)),
-          Dict(:alg=>TRBDF2(autodiff=false)),
-          Dict(:alg=>QNDF(autodiff=false)),
-          Dict(:alg=>FBDF(autodiff=false)),
-          Dict(:alg=>CVODE_BDF()),
-          Dict(:alg=>CVODE_BDF(linear_solver=:LapackDense)),
-          #Dict(:alg=>rodas()),
-          #Dict(:alg=>radau()),
-          #Dict(:alg=>Rodas4(autodiff=false)),
-          #Dict(:alg=>Rodas5(autodiff=false)),
-          Dict(:alg=>KenCarp4(autodiff=false)),
-          Dict(:alg=>KenCarp47(autodiff=false)),
-          #Dict(:alg=>RadauIIA5(autodiff=false)),
-          #Dict(:alg=>lsoda()),
+          Dict(:alg=>lsoda(), :prob_choice => 1),
+          Dict(:alg=>CVODE_BDF(linear_solver=:GMRES,prec=precilu_fceri,psetup=psetupilu_fceri,prec_side=1), :prob_choice => 2),
+          Dict(:alg=>TRBDF2(linsolve=KrylovJL_GMRES(),autodiff=false,precs=incompletelu_fceri,concrete_jac=true), :prob_choice => 2),
+          Dict(:alg=>QNDF(linsolve=KrylovJL_GMRES(),autodiff=false,precs=incompletelu_fceri,concrete_jac=true), :prob_choice => 2),
+          Dict(:alg=>FBDF(linsolve=KrylovJL_GMRES(),autodiff=false,precs=incompletelu_fceri,concrete_jac=true), :prob_choice => 2),
+          Dict(:alg=>KenCarp4(linsolve=KrylovJL_GMRES(),autodiff=false,precs=incompletelu_fceri,concrete_jac=true), :prob_choice => 2),
+          Dict(:alg=>Rosenbrock23(linsolve=KrylovJL_GMRES(),autodiff=false,precs=incompletelu_fceri,concrete_jac=true), :prob_choice => 2)

```

Comment:
Why Rosenbrock23? Sounds like Rodas5P would be better off here? Tolerances are too low.

Reply:
This is missing any usage of the sparse factorization methods. That seems like a big omission.

Reply:
By sparse factorisation method do you mean linsolver? 

Here I took the algorithms we used in the tests (with the best option for each). I am happy to remove/add some if you tell me what to use where.

Reply:
Yes. The bruss methods are a good template to look at: https://benchmarks.sciml.ai/stable/StiffODE/Bruss/ . On CVODE, show factorization vs GMRES, with and without preconditioners, with different problem setups. Then do that with one of the good Julia methods. Then take what is performing best and pit them against each other. Basically, a dive into the setup of the method, and then comparison between the good ones. That then paints the whole picture.